Review

Markezon opened the pull request.

Markezon committed:
add new confg

Markezon committed:
-настроен роутинг
-работают запросы к серверу
-настроен запрос авторизации
Теперь пользователь может:
зарегистрироваться,
авторизоваться,
создать заказ,
увидеть его в своём профиле.

Markezon committed:
delete env.example

ArslanMustafin commented on src/components/center-with-title/center-with-title.tsx:
```diff
@@ -0,0 +1,23 @@
+  const location = useLocation();
+  const [titleStyle, setTitleStyle] = useState('text_type_main-large');
+
+  useEffect(() => {
+    if (/feed|profile/i.test(location.pathname)) {
+      setTitleStyle('text_type_digits-default');
+    }
+  });
```
**Надо исправить:**

а точно ли есть в этом смысл? вы ведь используете этот компонент только на этих страницах, не совсем понимаю пользу условия.

Лучше не использовать `useEffect` для таких целей. Используйте `useMemo` чтобы получить нужный стиль, а регулярку можно заменить на `useMatch` https://reactrouter.com/en/main/start/overview#active-links

ArslanMustafin commented on src/components/modal/modal.tsx:
```diff
@@ -1,12 +1,22 @@
+  const location = useLocation();
+  const [titleStyle, setTitleStyle] = useState('text_type_main-large');
+
+  useEffect(() => {
+    if (/feed|profile/i.test(location.pathname)) {
+      setTitleStyle('text_type_digits-default');
+    }
+  });
```
**Надо исправить:**

аналогичная проблема что и в компоненте `Center`. Лучше не использовать `useEffect` для таких целей. Используйте `useMemo` чтобы получить нужный стиль, а регулярку можно заменить на `useMatch` https://reactrouter.com/en/main/start/overview#active-links

ArslanMustafin commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,25 @@
+  forAuthorized: boolean;
+};
+
+export const ProtectedRoute = ({
+  forAuthorized = false
+}: ProtectedRouteProps) => {
+  const location = useLocation();
+  const isAuthorized = useSelector(isAuthorizedSelector);
+  const from = location.state?.from || '/';
+
+  if (!forAuthorized && isAuthorized) {
+    return <Navigate to={from} />;
+  }
+
+  if (forAuthorized && !isAuthorized) {
+    return <Navigate to='/login' state={{ from: location }} />;
+  }
+
+  return <Outlet />;
+};
```
**Надо исправить:**

У вас мелькают защищённые страницы, при перезагрузке.
Вы должны тут просто дождаться получение пользователя и проверить авторизован он или нет.

Оптимальный пример компонента ProtectRoute:

```js
export const ProtectedRoute = ({ onlyUnAuth = false, children }) => {
  const isAuthChecked = useSelector((state) => state.user.isAuthChecked);
  const user = useSelector((state) => state.user.data);
  const location = useLocation();

  if (!isAuthChecked) {
    return <Preloader />;
  }

  if (onlyUnAuth && user) {
    const { from } = location.state || { from: { pathname: "/" } };
    return <Navigate to={from} />;
  }

  if (!onlyUnAuth && !user) {
    return (
      <Navigate
        to="/login"
        state={{ from: location }}
      />
    );
  }

  return children;
};

// использование

<Route
  path="/register"
  element={
    <ProtectedRoute onlyUnAuth={true}>
      <Register />
    </ProtectedRoute>
  }
/>

<Route
  path="/profile"
  element={
    <ProtectedRoute>
      <Profile />
    </ProtectedRoute>
  }
/>
```

Вам не хватает вот этой проверки, поэтому у вас и мелькают защищённые страницы.

```js
if (!isAuthChecked) {
  return <Preloader />;
}
```

`isAuthChecked` - это просто флаг, который показывает, выполнился ли запрос на получения пользователя или ещё нет. Важен сам процесс, если он завершён, то флаг будет `true` (не зависит от результат, т.к. он будет `true` в любом случае, даже если будет ошибка), если он ещё не произошёл, то будет `false`.

ArslanMustafin commented on src/services/slices/constructorSlice.ts:
```diff
@@ -0,0 +1,114 @@
+      } else {
+        state.constructorItems.ingredients.push({
+          ...action.payload,
+          id: nanoid()
```
**Надо исправить:**

Редьюсеры — чистые функции, в них не должно быть генерации каких-нибудь айдишников.
Для этого есть prepare. <https://redux-toolkit.js.org/api/createslice#customizing-generated-action-creators>

```js
//пример:

addToConstructor: {
  reducer: (state, { payload }: PayloadAction<TConstructorIngredient>) => {
    if (payload.type === 'bun') {
      state.bun = payload;
    } else {
      state.ingredients.push(payload);
    }
  },
  prepare: (ingredient: TIngredient) => ({
    payload: { ...ingredient, id: uuid() }
  })
},

```

ArslanMustafin commented on src/services/slices/userSlice.ts:
```diff
@@ -0,0 +1,184 @@
+        state.error = null;
+        state.user = payload.user;
+        state.isAuthorized = true;
+        setCookie('accessToken', payload.accessToken);
+        localStorage.setItem('refreshToken', payload.refreshToken);
```
**Надо исправить:**

здесь и далее.

Редьюсеры — чистые функции, в них не должно быть работы с localStorage и т.п. Для этого есть экшены. Можете там дождаться выполнение запроса) и поставить в куки и localStorage нужные вам данные.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа. 🔥

### **Вы молодец!**

- Сборка и запуск проекта выполняются без ошибок;
- Есть модальные окна;
- Данные получают с сервера;
- Есть регистрация и авторизация;

### **Но есть несколько замечаний, которые нужно исправить:**

- <https://disk.yandex.com/i/N_O9ROSupCqKrg> - В шапке сайта должна отображаться активной та ссылка на маршрут, где сейчас находится пользователь. Ссылка остаётся активной, если пользователь попадает на вложенные динамические маршруты.

- Отметил замечания в коде;

Markezon committed:
changes in:
src/components/center-with-title/center-with-title.tsx
src/components/modal/modal.tsx
src/components/protected-route/protected-route.tsx
src/services/slices/constructorSlice.ts
src/services/slices/userSlice.ts

Markezon committed:
В шапке сайта отображается активной та ссылка, где сейчас находится пользователь.

ArslanMustafin commented on src/components/ui/app-header/app-header.tsx:
```diff
@@ -1,49 +1,80 @@
+            <Link
+              className={clsx(
+                styles.link,
+                currentLocation === '/'
+                  ? [styles.link_active, styles.link]
+                  : styles.link
```
**Можно лучше:**

здесь и далее.

тут тоже лучше использовать `useMatch`)

ArslanMustafin reviewed (approved):
## Code Review

Здравствуйте!

Отличная работа. 🔥

### **Вы молодец! Работа принята, удачи в дальнейшем обучении!**

- Все замечания были исправлены;

Markezon committed:
jest+cypress config

Markezon committed:
Merge branch 'review' of github.com:Markezon/stellar-burger into review

Markezon committed:
src

Markezon committed:
readd files

Markezon committed:
Merge branch 'review' of github.com:Markezon/stellar-burger into review

Markezon committed:
+

Markezon committed:
add

Markezon committed:
Create constructor.cy.tsx

Markezon committed:
constructor.cy.tsx

Markezon reviewed (commented):
commit

Markezon committed:
add files

readme

Markezon reviewed (commented):
+

Markezon committed:
Add files via upload

Markezon committed:
Add files via upload

Markezon committed:
readme

Markezon committed:
add files

Markezon committed:
add

Markezon committed:
Create .env

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* Все юнит-тесты (38 шт) проходят успешно
* Все cypress-тесты проходят успешно
* Отлично, что знаете про `alias` (переменные с `@`)
* Код тестов аккуратный и понятный
* Отлично, что базовый урл вынесен в конфиг



Поздравляю! Ваша работа принята.


Критических замечаний нет

Вы отлично потрудились.

Удачного дальнейшего обучения.
